File: isucon14/webapp/python/app/owner_handlers.py
from collections import defaultdict
from collections.abc import MutableMapping
from datetime import datetime
from http import HTTPStatus
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Response
from pydantic import BaseModel
from sqlalchemy import text
from ulid import ULID

from .middlewares import owner_auth_middleware
from .models import Chair, Owner, Ride
from .sql import engine
from .utils import (
    datetime_fromtimestamp_millis,
    secure_random_str,
    sum_sales,
    timestamp_millis,
)

router = APIRouter(prefix="/api/owner")


class OwnerPostOwnersRequest(BaseModel):
    name: str


class OwnerPostOwnersResponse(BaseModel):
    id: str
    chair_register_token: str


@router.post("/owners", status_code=HTTPStatus.CREATED)
def owner_post_owners(
    req: OwnerPostOwnersRequest, response: Response
) -> OwnerPostOwnersResponse:
    if req.name == "":
        raise HTTPException(
            status_code=HTTPStatus.BAD_REQUEST,
            detail="some of required fields(name) are empty",
        )

    owner_id = str(ULID())
    access_token = secure_random_str(32)
    chair_register_token = secure_random_str(32)

    with engine.begin() as conn:
        conn.execute(
            text(
                "INSERT INTO owners (id, name, access_token, chair_register_token) VALUES (:id, :name, :access_token, :chair_register_token)"
            ),
            {
                "id": owner_id,
                "name": req.name,
                "access_token": access_token,
                "chair_register_token": chair_register_token,
            },
        )

    response.set_cookie(path="/", key="owner_session", value=access_token)

    return OwnerPostOwnersResponse(
        id=owner_id, chair_register_token=chair_register_token
    )


class ChairSales(BaseModel):
    id: str
    name: str
    sales: int


class ModelSales(BaseModel):
    model: str
    sales: int


class OwnerGetSalesResponse(BaseModel):
    total_sales: int
    chairs: list[ChairSales]
    models: list[ModelSales]


@router.get("/sales")
def owner_get_sales(
    owner: Annotated[Owner, Depends(owner_auth_middleware)],
    since: int | None = None,
    until: int | None = None,
) -> OwnerGetSalesResponse:
    if since is None:
        since_dt = datetime_fromtimestamp_millis(0)
    else:
        since_dt = datetime_fromtimestamp_millis(since)

    if until is None:
        until_dt = datetime(9999, 12, 31, 23, 59, 59)
    else:
        until_dt = datetime_fromtimestamp_millis(until)

    with engine.begin() as conn:
        rows = conn.execute(
            text("SELECT * FROM chairs WHERE owner_id = :owner_id"),
            {"owner_id": owner.id},
        ).fetchall()
        chairs = [Chair.model_validate(r) for r in rows]

        res = OwnerGetSalesResponse(total_sales=0, chairs=[], models=[])
        model_sales_by_model: MutableMapping[str, int] = defaultdict(int)
        for chair in chairs:
            rows = conn.execute(
                text(
                    "SELECT rides.* FROM rides JOIN ride_statuses ON rides.id = ride_statuses.ride_id WHERE chair_id = :chair_id AND status = 'COMPLETED' AND updated_at BETWEEN :since AND :until + INTERVAL 999 MICROSECOND"
                ),
                {
                    "chair_id": chair.id,
                    "since": since_dt,
                    "until": until_dt,
                },
            ).fetchall()
            rides = [Ride.model_validate(r) for r in rows]

            chair_sales = sum_sales(rides)

            res.total_sales += chair_sales
            res.chairs.append(
                ChairSales(id=chair.id, name=chair.name, sales=chair_sales)
            )
            model_sales_by_model[chair.model] += chair_sales

        model_sales = []
        for model, sales in model_sales_by_model.items():
            model_sales.append(ModelSales(model=model, sales=sales))

        res.models = model_sales

        return res


class ChairWithDetail(BaseModel):
    id: str
    owner_id: str
    name: str
    access_token: str
    model: str
    is_active: bool
    created_at: datetime
    updated_at: datetime
    total_distance: int
    total_distance_updated_at: datetime | None = None


class OwnerGetChairResponseChair(BaseModel):
    id: str
    name: str
    model: str
    active: bool
    registered_at: int
    total_distance: int
    total_distance_updated_at: int | None = None


class OwnerGetChairResponse(BaseModel):
    chairs: list[OwnerGetChairResponseChair]


@router.get(
    "/chairs",
    status_code=HTTPStatus.OK,
    response_model_exclude_none=True,
)
def owner_get_chairs(
    owner: Annotated[Owner, Depends(owner_auth_middleware)],
) -> OwnerGetChairResponse:
    with engine.begin() as conn:
        rows = conn.execute(
            text(
                """
                SELECT id,
                       owner_id,
                       name,
                       access_token,
                       model,
                       is_active,
                       created_at,
                       updated_at,
                       IFNULL(total_distance, 0) AS total_distance,
                       total_distance_updated_at
                FROM chairs
                       LEFT JOIN (SELECT chair_id,
                                          SUM(IFNULL(distance, 0)) AS total_distance,
                                          MAX(created_at)          AS total_distance_updated_at
                                   FROM (SELECT chair_id,
                                                created_at,
                                                ABS(latitude - LAG(latitude) OVER (PARTITION BY chair_id ORDER BY created_at)) +
                                                ABS(longitude - LAG(longitude) OVER (PARTITION BY chair_id ORDER BY created_at)) AS distance
                                         FROM chair_locations) tmp
                                   GROUP BY chair_id) distance_table ON distance_table.chair_id = chairs.id
                WHERE owner_id = :owner_id
        """
            ),
            {"owner_id": owner.id},
        )
        chairs = [ChairWithDetail.model_validate(r) for r in rows.mappings()]

    res = OwnerGetChairResponse(chairs=[])
    for chair in chairs:
        c = OwnerGetChairResponseChair(
            id=chair.id,
            name=chair.name,
            model=chair.model,
            active=chair.is_active,
            registered_at=timestamp_millis(chair.created_at),
            total_distance=chair.total_distance,
            total_distance_updated_at=None,
        )
        if chair.total_distance_updated_at is not None:
            t = timestamp_millis(chair.total_distance_updated_at)
            c.total_distance_updated_at = t
        res.chairs.append(c)

    return res

File: isucon14/webapp/nodejs/src/owner_handlers.ts
import type { Context } from "hono";
import type { Environment } from "./types/hono.js";
import { secureRandomStr } from "./utils/random.js";
import { setCookie } from "hono/cookie";
import type { RowDataPacket } from "mysql2";
import type { Chair, Ride } from "./types/models.js";
import { calculateSale } from "./common.js";
import { ulid } from "ulid";

export const ownerPostOwners = async (ctx: Context<Environment>) => {
  const reqJson = await ctx.req.json<{ name: string }>();
  const { name } = reqJson;
  if (!name) {
    return ctx.text("some of required fields(name) are empty", 400);
  }
  const ownerId = ulid();
  const accessToken = secureRandomStr(32);
  const chairRegisterToken = secureRandomStr(32);
  await ctx.var.dbConn.query(
    "INSERT INTO owners (id, name, access_token, chair_register_token) VALUES (?, ?, ?, ?)",
    [ownerId, name, accessToken, chairRegisterToken],
  );

  setCookie(ctx, "owner_session", accessToken, { path: "/" });

  return ctx.json(
    { id: ownerId, chair_register_token: chairRegisterToken },
    201,
  );
};

export const ownerGetSales = async (ctx: Context<Environment>) => {
  const since = new Date(0);
  const until = new Date("9999-12-31T23:59:59Z");
  if (ctx.req.query("since")) {
    since.setTime(Number(ctx.req.query("since")));
  }
  if (ctx.req.query("until")) {
    until.setTime(Number(ctx.req.query("until")));
  }

  const owner = ctx.var.owner;
  await ctx.var.dbConn.beginTransaction();
  try {
    const [chairs] = await ctx.var.dbConn.query<Array<Chair & RowDataPacket>>(
      "SELECT * FROM chairs WHERE owner_id = ?",
      [owner.id],
    );

    let totalSales = 0;
    const chairSales = [];
    const modelSalesByModel: { [key: string]: number } = {};
    for (const chair of chairs) {
      const [rides] = await ctx.var.dbConn.query<Array<Ride & RowDataPacket>>(
        "SELECT rides.* FROM rides JOIN ride_statuses ON rides.id = ride_statuses.ride_id WHERE chair_id = ? AND status = 'COMPLETED' AND updated_at BETWEEN ? AND ? + INTERVAL 999 MICROSECOND",
        [chair.id, since, until],
      );
      const sales = sumSales(rides);
      totalSales += sales;
      chairSales.push({ id: chair.id, name: chair.name, sales });
      modelSalesByModel[chair.model] =
        (modelSalesByModel[chair.model] ?? 0) + sales;
    }

    const models = Object.entries(modelSalesByModel).map(([model, sales]) => ({
      model,
      sales,
    }));

    return ctx.json({
      total_sales: totalSales,
      chairs: chairSales,
      models,
    });
  } catch (e) {
    await ctx.var.dbConn.rollback();
    return ctx.text(`${e}`, 500);
  }
};

type ChairWithDetail = {
  id: string;
  owner_id: string;
  name: string;
  access_token: string;
  model: string;
  is_active: boolean;
  created_at: Date;
  updated_at: Date;
  total_distance: number;
  total_distance_updated_at: Date | null;
};

type OwnerGetChairsResponseChair = {
  id: string;
  name: string;
  model: string;
  active: boolean;
  registered_at: number;
  total_distance: number;
  total_distance_updated_at?: number;
};

export const ownerGetChairs = async (ctx: Context<Environment>) => {
  const owner = ctx.var.owner;
  const [chairs] = await ctx.var.dbConn.query<
    Array<ChairWithDetail & RowDataPacket>
  >(
    `SELECT id,
       owner_id,
       name,
       access_token,
       model,
       is_active,
       created_at,
       updated_at,
       IFNULL(total_distance, 0) AS total_distance,
       total_distance_updated_at
FROM chairs
       LEFT JOIN (SELECT chair_id,
                          SUM(IFNULL(distance, 0)) AS total_distance,
                          MAX(created_at)          AS total_distance_updated_at
                   FROM (SELECT chair_id,
                                created_at,
                                ABS(latitude - LAG(latitude) OVER (PARTITION BY chair_id ORDER BY created_at)) +
                                ABS(longitude - LAG(longitude) OVER (PARTITION BY chair_id ORDER BY created_at)) AS distance
                         FROM chair_locations) tmp
                   GROUP BY chair_id) distance_table ON distance_table.chair_id = chairs.id
WHERE owner_id = ?`,
    [owner.id],
  );

  const chairResponse = chairs.map((chair) => {
    const c: OwnerGetChairsResponseChair = {
      id: chair.id,
      name: chair.name,
      model: chair.model,
      active: !!chair.is_active,
      registered_at: chair.created_at.getTime(),
      total_distance: Number(chair.total_distance),
    };
    if (chair.total_distance_updated_at) {
      c.total_distance_updated_at = chair.total_distance_updated_at.getTime();
    }
    return c;
  });

  return ctx.json({ chairs: chairResponse });
};

function sumSales(rides: Ride[]) {
  return rides.reduce((acc, ride) => acc + calculateSale(ride), 0);
}
